extensions/amp-story/1.0/amp-story-reaction.js

```diff
@@ -110,45 +113,74 @@ export class AmpStoryReaction extends AMP.BaseElement {
     /** @protected {!Element} */
     this.rootEl_;
 
-    /** @protected {?string} */
-    this.reactionId_ = null;
+    /** @protected {string} */
+    this.reactionId_;
 
     /** @protected {!./amp-story-request-service.AmpStoryRequestService} */
     this.requestService_ = getRequestService(this.win, this.element);
 
     /** @protected {?Promise<?ReactionResponseType|?JsonObject|undefined>} */
     this.responseDataPromise_ = null;
 
-    /** @protected {?ReactionResponseType} */
-    this.responseData_ = null;
+    /** @protected {?Array<ReactionOptionType>} */
+    this.responseData_;
 
     /** @const @protected {!./amp-story-store-service.AmpStoryStoreService} */
     this.storeService_ = getStoreService(this.win);
 
     /** @const @protected {!./variable-service.AmpStoryVariableService} */
     this.variableService_ = getVariableService(this.win);
+
+    /** @protected {!Array<!Element>} */
+    this.optionElements_ = [];
+  }
+
+  /**
+   * Get the root element
+   * @visibleForTesting
+   * @return {Element}
+   */
+  getRootElement() {
+    return this.rootEl_;
+  }
+
+  /**
+   * Get the options
+   * @visibleForTesting
+   * @return {Array<Element>}
+   */
+  getOptionElements() {
+    return this.optionElements_;
   }
 
   /** @override */
   buildCallback() {
-    this.buildComponent(this.element);
+    this.rootEl_ = this.buildComponent(this.element);
     this.element.classList.add('i-amphtml-story-reaction');
     this.adjustGridLayer_();
     this.initializeListeners_();
     createShadowRootWithStyle(this.element, this.rootEl_, CSS);
   }
 
   /**
-   * Generates the template in rootEl_ and fills up with options.
+   * Generates the template in this.rootEl_ and fills up with options.
    * @param {!Element} unusedElement
    * @protected @abstract
+   * @return {!Element}
    */
   buildComponent(unusedElement) {
     // Subclass must override.
   }
 
   /** @override */
   layoutCallback() {
+    const pageId = closest(dev().assertElement(this.element), (el) => {
+      return el.tagName.toLowerCase() === 'amp-story-page';
+    }).getAttribute('id');
+    this.reactionId_ = `CANONICAL_URL+${pageId}`;
+    if (this.element.hasAttribute('id')) {
+      this.reactionId_ += '+' + this.element.getAttribute('id');
```

Comment:
This comes from the conversation of how to control the number of reactions per page, because if more than one is needed, then we need the element ID as well. In the case that we can enforce the "one per page", we might not need the quiz id, but we should talk about it and come up with a plan, since right now we really can't enforce anything.